devdoc/channel.md

```diff
@@ -0,0 +1,267 @@
+# `channel` requirements
+
+## Overview
+
+`channel` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers of data call `channel_pull` to register a `PULL_CALLBACK` function to be called when data becomes available.
+
+Producers of data call `channel_push` to notify the channel that data is available.
+
+`channel_pull` and `channel_push` can be called from different threads in any order.
+
+## `channel` state
+
+The state of the `channel` module consists of two components:
+- EPOCH: This is a monotonically increasing integer that is incremented every time an operation is completed or cancelled.
+- Internal state: This is an enum that can have the following values:
+  - `CHANNEL_READY`: This is the default state. In this state, the module accepts calls to `channel_pull` and `channel_push`
+  - `CHANNEL_PUSHING`: This is a temporary state when `channel_push` is executing. It causes concurrent calls to `channel_pull` to wait.
+  - `CHANNEL_PUSHED`: This is the state when `channel_push` has completed execution. A subsequent call to `channel_pull` will result in the callbacks being called synchronously.
+  - `CHANNEL_PULLING`: This is a temporary state when `channel_pull` is executing. It causes concurrent calls to `channel_push` to wait.
+  - `CHANNEL_PULLED`: This is the state when `channel_pull` has completed execution. A subsequent call to `channel_push` will result in the callback being called synchronously.
+
+Both components are stored in a single `volatile_atomic` variable that is manipulated using `interlocked` APIs.
+
+```mermaid
+stateDiagram-v2
+    B: EPOCH=x,STATE=CHANNEL_READY
+    N: EPOCH=x,STATE=CHANNEL_PUSHED
```

Comment:
>x [](http://example.com/codeflow?start=13&length=1)

seeing in this diagram as the state doesn't increase until it returns to "CHANNEL_READY" I am starting to think that the "operation" from line 16 is a complete cycle. 

oO